model/task/task.go

```diff
@@ -775,6 +778,39 @@ func (t *Task) AllDependenciesSatisfied(cache map[string]Task) (bool, error) {
 	return true, nil
 }
 
+// MarkDependenciesFinished updates all direct dependencies on this task to
+// cache whether or not this task has finished running.
+func (t *Task) MarkDependenciesFinished(finished bool) error {
```

Comment:
Is there ever a case where we would need to mark this as `false`?

Reply:
Yes, it gets marked false (i.e. task is not finished) when the task is restarted to indicate that the dependency is no longer finished running.

It would be preferable to derive the finished state from the task's in-memory status (e.g. "completed", "failed") rather than pass it in as a parameter, but I'm not confident that `RestartVersion`, `RestartBuild`, and `RestartAllBuildTasks` (which call `MarkTasksReset`) actually reset execution tasks to the "undispatched" status. It seems like [`MarkTasksReset` only resets the status of display tasks](https://github.com/evergreen-ci/evergreen/blob/d08da70677174ecc33f9b70e2d0e4d708269e129/model/task_lifecycle.go#L1253-L1261), so when I wrote a test for `RestartVersions` it failed to mark the task's dependencies as unfinished because the task was never reset. Hence, I had to manually tell it to unset the dependencies.

I'm not entirely confident that the restart logic is correct for builds/versions, but I don't really want to mess with the existing logic to find out. There's [a ticket](https://jira.mongodb.org/browse/EVG-16480) to investigate the confusion I mentioned above.